hudi-client/hudi-spark-client/src/main/scala/org/apache/spark/sql/BucketPartitionUtils.scala

```diff
@@ -0,0 +1,54 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql
+
+import org.apache.hudi.common.model.HoodieRecord
+import org.apache.hudi.index.bucket.BucketIdentifier
+import org.apache.spark.Partitioner
+import org.apache.spark.sql.catalyst.InternalRow
+
+object BucketPartitionUtils {
+  def createDataFrame(df: DataFrame, indexKeyFields: String, bucketNum: Int, partitionNum: Int): DataFrame = {
+    def getPartitionKeyExtractor(): InternalRow => Int = row => {
+      val kb = BucketIdentifier
+        .getBucketId(row.getString(HoodieRecord.HOODIE_META_COLUMNS.indexOf(HoodieRecord.RECORD_KEY_METADATA_FIELD)), indexKeyFields, bucketNum)
+      val partition = row.getString(HoodieRecord.HOODIE_META_COLUMNS.indexOf(HoodieRecord.PARTITION_PATH_METADATA_FIELD))
+      if (partition == null || partition.trim.isEmpty) {
+        kb
+      } else {
+        val pw = (partition.hashCode & Int.MaxValue) % partitionNum
+        BucketIdentifier.mod(kb + pw, partitionNum)
+      }
+    }
+
+    // use internalRow to avoid extra convert.
+    val reRdd = df.queryExecution.toRdd.mapPartitions { iter => {
+      val getPartitionKey = getPartitionKeyExtractor()
+      iter.map { row => (getPartitionKey(row), row) }
+    }
+    }.partitionBy(new Partitioner {
+      override def numPartitions: Int = partitionNum
+
+      override def getPartition(key: Any): Int = {
+        key.asInstanceOf[Int]
```

Comment:
The partition does not sort by partition path actually.

Reply:
@danny0405 sort by partition path added

Reply:
I'm not very familiar with Spark but it seems the output partitions of Spark is a fixed number, if we key by `partiton path` + `bucket id`, there could be still multiple partitions in one Spark shard, the sort was to avoid to hold too many file handles when the partitions number are huge.

Reply:
the current PartitionKey(partition path + bucket id), one spark shuffle split  reader may contain multiple partitions, but it with diff bucketId to create bucket file, I think it's normal.
for example：
partition key: a, b, c
bucketNumer 5, id maybe: 0,1,2,3,4
after customized partitioner, assum the parallelism is 10:
result one:
split 1: a-1, a-2, b-0
split 2: a-4, c-3, c-1
split 3: a-3
split4: a-0, b-1, b-2, b-3, b-4, c-0, c-2, c-4

and will use fanout writer to write every split. Partition may be tilted if the partitionedKey is unreasonable, but I think the number of files is fixed.

Reply:
> the current PartitionKey(partition path + bucket id), one spark shuffle split reader may contain multiple partitions, but it with diff bucketId to create bucket file, I think it's normal. for example： partition key: a, b, c bucketNumer 5, id maybe: 0,1,2,3,4 after customized partitioner, assum the parallelism is 10: result one: split 1: a-1, a-2, b-0 split 2: a-4, c-3, c-1 split 3: a-3 split4: a-0, b-1, b-2, b-3, b-4, c-0, c-2, c-4
> 
> and will use fanout writer to write every split. Partition may be tilted if the partitionedKey is unreasonable, but I think the number of files is fixed.

file is fixed，but sort can improve serialize speed and reduce write handlers switch to improve performance